Update and restructure documentation (#112)
eclipsesource/jsonforms2-website: src/pages/docs/react-integration.mdx

    },
    {
      "type": "Control",
      "scope": "#/properties/name"
    },
    {
      "type": "HorizontalLayout",
      "elements": [
        {
          "type": "Control",
          "scope": "#/properties/due_date"
        },
        {
          "type": "Control",
          "scope": "#/properties/rating"
        }
      ]
    },
    {
      "type": "Control",
      "scope": "#/properties/description",
      "options": {
        "multi": true
      }
    },
    {
      "type": "HorizontalLayout",
      "elements": [
        {
          "type": "Control",
          "scope": "#/properties/recurrence"
        },
        {
          "type": "Control",
          "scope": "#/properties/recurrence_interval",
          "rule": {
            "effect": "HIDE",
            "condition": {
              "type": "LEAF",
              "scope": "#/properties/recurrence",
              "expectedValue": "Never"
            }
          }
        }
      ]
    }
  ]
}
```

## `data`

The `data` prop represents an object containing the data to be rendered in the form.

<Note
  title='ATTENTION'
  body='Never update the data prop except when you intentionally want to set new data.'
/>

> The data given to JSON Forms should only be updated when necessary, for example, when clearing a form:

Reviewer: ```suggestion
JSON Forms maintains its data internally, treating `data` as the initial data. Therefore the data given to JSON Forms should only be updated when necessary because whenever JSON Forms receives new data it'll execute a validation and rerender pass.

You should only update the data object given to JSON Forms when you explicitly want to set new data from the 'outside', for example when you want to clear the form:
```

Reviewer: This is obsolete. The warning can be removed and all examples be updated so that they update the data. Actually this is now the more stable approach so we should encourage it.

```jsx
import React, { useState, useEffect } from 'react';
import { JsonForms } from '@jsonforms/react';
import schema from './schema.json';
import uischema from './uischema.json';
import {
  materialCells,
  materialRenderers,
} from '@jsonforms/material-renderers';
import { Button } from '@material-ui/core';

const data = {
  name: 'Send email to Adrian',
  description: 'Confirm if you have passed the subject\nHereby ...',
  done: true,
  recurrence: 'Daily',
  rating: 3,
};

const ClearFormExample = () => {
  const [displayDataAsString, setDisplayDataAsString] = useState('');
  const [jsonformsOutputData, setJsonformsOutputData] = useState < any > data;

>   const [jsonformsInputData, setJsonformsInputData] = useState < any > data;

Reviewer: I would keep the example free from Typescript, also parenthesis are missing ;)

  useEffect(() => {
    setDisplayDataAsString(JSON.stringify(jsonformsOutputData, null, 2));
  }, [jsonformsOutputData]);
  return (
    <div>
      <pre id='boundData'>{displayDataAsString}</pre>
      <JsonForms
        schema={schema}
        uischema={uischema}
        data={jsonformsInputData}
        renderers={materialRenderers}
        cells={materialCells}
        onChange={({ data }) => setJsonformsOutputData(data)}
      />
      <Button onClick={() => setJsonformsInputData({})} color='primary'>
        Clear form data
      </Button>
    </div>
  );
};

> ```

Reviewer: This example should be simplified. I'll come back when I have a suggestion ;)

Reviewer: The example can just update the data handed over so the initial data is only used for initializing the data state.

You should _never_ refeed JSON Forms with the data received from its `onChange` handler:

```jsx
const data = {
  name: 'Send email to Adrian',
  description: 'Confirm if you have passed the subject\nHereby ...',
  done: true,
  recurrence: 'Daily',
  rating: 3,
};

const DoNotDoThis = () => {

>   const [standaloneData, setStandaloneData] = useState < any > data;

Reviewer: Also remove Typescript and check parenthesis here.

  return (
    <JsonForms
      schema={schema}
      uischema={uischema}
      data={standaloneData}
      renderers={materialRenderers}
      cells={materialCells}
      onChange={({ data }) => {
        // Do not do this
        // we modify the standaloneData, which we also pass to the data prop
        setStandaloneData(data);
      }}
    />
  );
};
```

## `renderers`

With the `renderers` prop you can supply the renderers that should be used for rendering the form.
You can choose one of the [renderer sets](/docs/renderer-sets) already provided by JSON Forms or supply [your own renderers](/docs/custom-renderers).

## `cells`

> While the renderers supplied via the `renderers` prop are responsible for rendering the contents of the form,
> table controls use a special set of renderers for their cells, which can be provided via the `cells` prop.

Reviewer: ```suggestion
Cells are "simplistic" renderers which only render the actual input without displaying a label or validation messages. In the React Material renderer set cells are only used within the table control, as labels and validation are handled by the table renderer. In the React Vanilla renderer set cells are used both within tables but also as a basis for all other renderers, e.g. the integer renderer will also use an integer cell.
```

You can use the cell renderers provided by one of the available [renderer sets](/docs/renderer-sets) or supply [your own custom ones](/docs/custom-renderers).

## `onChange`

A callback which is called on each data change, containing the updated data and the validation result. JSON Forms will immediately produce
an event with the results of the initial validation, even before the inputs of the form are changed.

<Note
  title='ATTENTION'

>   body='You should never use the onChange handler to update the data prop, except when you intentionally want to set new data.'

Reviewer: ```suggestion
  body='You should not use the onChange handler to update the data given to JSON Forms itself'
```

/>

> Please refer to the information in the [data](#data) section above about correctly using the `onChange` handler to update the form data.

Reviewer: ```suggestion
```

## `config`

> You can configure some options available for all UI schema elements via the `config` prop:

Reviewer: ```suggestion
You can configure the default ui schema options for all UI schema elements via the `config` prop. The following options are supported by most of the provided renderers:
```

> - `restrict`: _boolean_ value specifying whether the number of characters should be restricted to 'maxLength' for inputs, if specified in the JSON schema

Reviewer: ```suggestion
- `restrict`: _boolean_ value specifying whether the number of characters should be restricted to 'maxLength' specified in the JSON schema
```

> - `trim`: _boolean_ value specifying whether the input width should be resized to 'maxLength', if specified in the JSON schema

Reviewer: `trim` indicates whether the control shall grab full width

> - `showUnfocusedDescription`: _boolean_ value specifying whether the input descriptions should hide when the input is not focused

Reviewer: ```suggestion
- `showUnfocusedDescription`: _boolean_ value specifying whether the input descriptions should be shown even when the input isn't focused
```

- `hideRequiredAsterisk`: _boolean_ value specifying whether the asterisks shown in labels for required fields should be hidden

The default config used by JSON Forms looks like this:

```
defaultConfigOptions = {
 restrict: false,
 trim: false,
 showUnfocusedDescription: false,
 hideRequiredAsterisk: false
}
```

## `uischemas`

> The `uischemas` prop can be used to register a list of UI schema and tester pairs that will be used for specific schema elements, e.g. for the detail of an element with detail.

Reviewer: ```suggestion
The `uischemas` prop can be used to register a list of UI schemas and corresponding testers that will be used whenever a 'detail' UI schema shall be rendered. This is useful when you need some kind of dynamic dispatching of uischemas.
```

> ## `readonly`

Reviewer: ```suggestion
## `readOnly`
```
